src/core/util/byte_stream.cc

```diff
@@ -100,6 +100,8 @@ void OutputByteStream::ProduceData(std::size_t amount) {
 }
 
 void OutputByteStream::WriteData(const std::uint8_t* data, std::size_t len) {
+  if (!len)
+    return;
```

Comment:
> not 100% sure why'd you want to do this or why you don't want to throw on 0 length or nullptr

Here is a sample use case : from [SSU RelayRequest spec](https://geti2p.net/spec/ssu), 

> the IP length is always 0 and the port is always 0, and the receiver should use the packet's source address and port

Or 

> Challenge is unimplemented, challenge size is always zero

IMHO, the packet builder should not care about these kind of rules. They should only be expressed in session.cc. To kinda explicitly say that it's the intended behavior, and not an oversight or whatever, i want to put something like this in session.cc

> packet.SetIPAddress(nullptr, 0);
> packet.SetChallenge(nullptr, 0);

But that would require code like 

> if(packet->GetIPAddressSize()) 
> WriteData(packet->GetIPAddress(), packet->GetIPAddressSize());
> if(packet->GetChallengeSize()) 
> WriteData(packet->GetChallenge(), packet->GetChallengeSize());

So what do you think about debug log ***if*** data null **and** len null, **otherwise** error log + throw  exception ?